test/integration-serial/serial_deploy_test.rb

```diff
@@ -235,6 +235,176 @@ def test_all_expected_statsd_metrics_emitted_with_essential_tags
+    ])
+  ensure
+    wait_for_all_crd_deletion
+  end
+
+  def test_cr_success_with_only_success_rollout_conditions
+    crd_result = deploy_fixtures("crd", subset: ["with_custom_conditions.yml"]) do |resource|
+      crd = resource["with_custom_conditions.yml"]["CustomResourceDefinition"].first
+      crd["metadata"]["annotations"][KubernetesDeploy::CustomResourceDefinition::ROLLOUT_CONDITIONS_ANNOTATION] = {
+        success_conditions: [{ path: "$.status.test_field", value: "success_value" }],
+      }.to_json
+    end
+    assert_deploy_success(crd_result)
+
+    success_conditions = {
+      "spec" => {},
+      "status" => {
+        "observedGeneration" => 1,
+        "test_field" => "success_value",
+        "condition" => "success_value",
+      },
+    }
+
+    result = deploy_fixtures("crd", subset: ["with_custom_conditions_cr.yml"]) do |resource|
+      cr = resource["with_custom_conditions_cr.yml"]["Customized"].first
+      cr.merge!(success_conditions)
+    end
+    assert_deploy_success(result)
+    assert_logs_match_all([
+      %r{Successfully deployed in .*: Customized\/with-customized-params},
+    ])
+  ensure
+    wait_for_all_crd_deletion
+  end
+
+  def test_cr_failure_with_arbitrary_rollout_conditions
+    assert_deploy_success(deploy_fixtures("crd", subset: ["with_custom_conditions.yml"]))
+    cr = load_fixtures("crd", ["with_custom_conditions_cr.yml"])
+    failure_conditions = {
+      "spec" => {},
+      "status" => {
+        "test_field" => "failure_value",
+        "error_msg" => "test error message jsonpath",
+        "observedGeneration" => 1,
+        "condition" => "failure_value",
+      },
+    }
+
+    result = deploy_fixtures("crd", subset: ["with_custom_conditions_cr.yml"]) do |resource|
+      cr = resource["with_custom_conditions_cr.yml"]["Customized"].first
+      cr.merge!(failure_conditions)
+    end
+    assert_deploy_failure(result)
+    assert_logs_match_all([
+      "test error message jsonpath",
+      "test custom error message",
+    ])
+  ensure
+    wait_for_all_crd_deletion
+  end
+
+  def test_deploying_cr_with_invalid_crd_conditions_fails
+    # Since CRDs are not always deployed along with their CRs and kubernetes-deploy is not the only way CRDs are
+    # deployed, we need to model the case where poorly configured rollout_conditions are present before deploying a CR
+    KubernetesDeploy::DeployTask.any_instance.expects(:validate_resources).returns(:true)
+    crd_result = deploy_fixtures("crd", subset: ["with_custom_conditions.yml"]) do |resource|
+      crd = resource["with_custom_conditions.yml"]["CustomResourceDefinition"].first
+      crd["metadata"]["annotations"].merge!(
+        KubernetesDeploy::CustomResourceDefinition::ROLLOUT_CONDITIONS_ANNOTATION => "blah"
+      )
+    end
+    assert_deploy_success(crd_result)
+    KubernetesDeploy::DeployTask.any_instance.unstub(:validate_resources)
+
+    cr_result = deploy_fixtures("crd", subset: ["with_custom_conditions_cr.yml"])
```

Comment:
As it stands this seems quite tricky to do and I'm not convinced it's the desired behaviour. Currently, we will print the error for each CR associated with a poorly specced CRD. I'm wondering how to get around this since
- All CRs contain a reference to their parent CRD
- In `validate_definition` of the CRD, `@validation_errors` will contain some errors.
- I'm not sure how/if to short-circuit this so that if the CRD is already invalidated we suppress the errors. We _could_ set `@validation_errors` to `nil` _if_ `@rollout_conditions_validated?`

Reply:
On further reflection I feel strongly that we should raise the error per CR. We should definitely make the error more explicit that the CR is failing because of the CRD as you suggest below.

Reply:
ok, I agree with you.